spectator-reg-metrics3/src/main/java/com/netflix/spectator/metrics3/MetricsRegistry.java

```diff
@@ -57,4 +54,42 @@ private String toMetricName(Id id) {
     final String name = toMetricName(id);
     return new MetricsTimer(clock(), id, impl.timer(name));
   }
+
+  @Override public <T extends Number> T gauge(Id id, T number) {
+    return gauge(id, number, (ToDoubleFunction<T>) value -> number.doubleValue());
+  }
+
+  @Override public <T> T gauge(Id id, T obj, ToDoubleFunction<T> f) {
+    final String name = toMetricName(id);
+
+    Gauge aggrGauge = this.impl.getGauges().get(name);
+
+    // Wasn't registered already
+    if (aggrGauge == null) {
+      aggrGauge = new MetricsGaugeAggr();
+      this.impl.register(name, aggrGauge);
```

Comment:
I would prefer not to lock, though I agree it would it is probably a rare case. Other options:

* Catch the exception from the duplicate registration and do another get.
* Maintain our own concurrent map of registered gauges. It looks like the `getGauges` call results in building a map of all gauges so this approach might also result in less temporary garbage getting created for each registration.